sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala

```diff
@@ -107,4 +109,85 @@ object LongExactNumeric extends LongIsIntegral with Ordering.LongOrdering {
   override def times(x: Long, y: Long): Long = Math.multiplyExact(x, y)
 
   override def negate(x: Long): Long = Math.negateExact(x)
+
+  override def toInt(x: Long): Int =
+    if (x == x.toInt) {
+      x.toInt
+    } else {
+      throw new ArithmeticException(s"Casting $x to Int causes overflow.")
+    }
+}
+
+object FloatExactNumeric extends FloatIsFractional with Ordering.FloatOrdering {
+  private def exceptionMessage(x: Float, dataType: String): String =
```

Comment:
nit: what about throwing the exception here?

Reply:
I thought abou it, but then it won't match the return type of `toInt` and `toLong`

Reply:
I am not sure I get your comment. If we create the exception in this method and return it, what does it change to `toLong` and `toInt`?

Reply:
one way to do it
```
def fail[T]: T = {
  throw ...
}
```

Reply:
OK, I see. We can write it like this:
```
  private def throwException(x: Float, dataType: String) =
    throw new ArithmeticException(s"Casting $x to $dataType causes overflow.")
```

We don't need to specify the returned type, which is `Nothing`.